includes/Core/Util/Entity_Factory.php

```diff
@@ -484,4 +497,49 @@ private static function prefix_title( $title, $prefix ) {
 			$title
 		);
 	}
+
+	/**
+	 * Gets the entity for the given URL, if available.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $url URL to determine the entity from.
+	 * @param Entity $entity The initial entity.
+	 * @return Entity The initial or new entity for the given URL.
+	 */
+	private static function maybe_convert_to_amp_entity( $url, $entity ) {
+		$url_parts    = wp_parse_url( $url );
+		$new_url_tail = '';
+
+		wp_parse_str( $url_parts['query'], $url_query_params );
+
+		// check if the $url has amp query param.
+		if ( array_key_exists( 'amp', $url_query_params ) ) {
+			$new_url_tail = '?amp=1';
+		}
+
+		// check if the $url has `/amp` in path.
+		if ( '/amp' === substr( untrailingslashit( $url_parts['path'] ), -4 ) ) { // -strlen('/amp') is -4
+			$new_url_tail = '/amp';
+		}
+
+		if ( empty( $new_url_tail ) ) {
+			return $entity;
+		}
+
+		$new_url = $entity->get_url();
+		$new_url = $new_url . $new_url_tail;
```

Comment:
We can't just concatenate the entity URL and the new tail because the existing entity URL may already have query parameters which will lead to something like this: `http://example.com/?p=12345?amp=1`.